# Question 5 SVM Programming Assignment

## Preprocessing Step(Preparing the Data)

In [37]:
#make the necessary imports
import numpy as np
import pandas as pd
from sklearn.svm import SVC #importing sklearn's svm module
from sklearn.metrics import accuracy_score #importing accuracy_score func from sklearn's metrics module

### Note: As the Testing labels were not present for this problem, so I have considered validation data for testing the model accuracy

In [38]:
fil= open("gisette_train.data") #opening the file
X_train=[]
for row in fil.readlines(): #reading the file line by line
    X_train.append((row.strip()).split()) #to get the values tuple by tuple without any whitespaces
fil.close()
#after this train will contain all the data in form of a numpy array

#repeating the same thing for test and valid samples as well

fil= open("gisette_train.labels")
y_train=[]
for row in fil.readlines():
    y_train.append((row.strip()).split())
fil.close()

fil= open("gisette_valid.data")
X_test=[]
for row in fil.readlines():
    X_test.append((row.strip()).split())
fil.close()

fil= open("gisette_valid.labels")
y_test=[]
for row in fil.readlines():
    y_test.append((row.strip()).split())
fil.close()

#some type conversions
X_train=np.array(X_train).astype(int)
y_train= np.array(y_train).astype(int)[:,0]

#some type conversions
X_test=np.array(X_test).astype(int)
y_test= np.array(y_test).astype(int)[:,0]

In [39]:
#converting from numpy array to pandas dataframe
X = pd.DataFrame(X_train)
y = y_train
val_train = pd.DataFrame(X_test)
val_test = y_test

# 5. (a)

## Using Linear Kernel

In [40]:
#we will be dealing with linear kernel here
svc=SVC(kernel='linear')

In [41]:
#training the linear kernel
svc.fit(X,y)

preds_train=svc.predict(X) #making predictions
print("Training Error: {}".format(1-accuracy_score(y,preds_train)))

Training Error: 0.0


#### Training Error: 0.0

In [42]:
preds_test=svc.predict(val_train) #making predictions
print("Testing Error: {}".format(1-accuracy_score(val_test,preds_test)))

Testing Error: 0.02400000000000002


#### Testing Error: 0.02400000000000002

In [43]:
print("Classwise count of support vectors: {}".format(svc.n_support_))
#for getting the number of suppport vectors for each class
print("\n")
print("Total number of support vectors: {}".format(len(svc.support_vectors_)))
#for getting the total count of the support vectors

Classwise count of support vectors: [542 542]


Total number of support vectors: 1084


#### Classwise count of support vectors: [542 542]
#### Total number of support vectors: 1084

# 5. (b)

## Using rbf kernel

In [25]:
rbf_svc = SVC(kernel='rbf', gamma=0.001)
#using the rbf kernel with gamma=0.001

In [26]:
rbf_svc.fit(X,y) #training the rbf kernel
preds_train=rbf_svc.predict(X) #making predictions
print("Training Error: {}".format(1-accuracy_score(y,preds_train)))

Training Error: 0.0


#### Training Error: 0.0

In [27]:
preds_test=rbf_svc.predict(val_train) #making predictions
print("Testing Error: {}".format(1-accuracy_score(val_test,preds_test)))

Testing Error: 0.5


#### Testing Error: 0.5

In [28]:
print("Classwise count of support vectors: {}".format(rbf_svc.n_support_))
#for getting the number of suppport vectors for each class
print("\n")
print("Total number of support vectors: {}".format(len(rbf_svc.support_vectors_)))
#for getting the total count of the support vectors

Classwise count of support vectors: [3000 3000]


Total number of support vectors: 6000


#### Classwise count of support vectors: [3000 3000]
#### Total number of support vectors: 6000

## Using Polynomial kernel

In [33]:
poly_svc = SVC(kernel = 'poly', degree = 2, coef0= 1)
#degree=2 and coef0=1 because we want a poly kernel of the form (1+x^T.x)^2

In [34]:
poly_svc.fit(X,y) #training the poly kernel model
preds_train=poly_svc.predict(X) #making predictions
print("Training Error: {}".format(1-accuracy_score(y,preds_train)))

Training Error: 0.0004999999999999449


#### Training Error: 0.0004999999999999449

In [35]:
preds_test=poly_svc.predict(val_train) #making predictions
print("Testing Error: {}".format(1-accuracy_score(val_test,preds_test)))

Testing Error: 0.020000000000000018


#### Testing Error: 0.020000000000000018

In [36]:
print("Classwise count of support vectors: {}".format(poly_svc.n_support_))
#for getting the number of suppport vectors for each class
print("\n")
print("Total number of support vectors: {}".format(len(poly_svc.support_vectors_)))
#for getting the total count of the support vectors

Classwise count of support vectors: [641 691]


Total number of support vectors: 1332


#### Classwise count of support vectors: [641 691]
#### Total number of support vectors: 1332

### linear and rbf both yielded the lowest training error of 0